bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeHandler.java

```diff
@@ -0,0 +1,549 @@
+    }
+
+    private @Nullable State updateMode() throws GreeException {
+        if (device.hasStatusValChanged(GREE_PROP_MODE)) {
+            int mode = device.getIntStatusVal(GREE_PROP_MODE);
+            String modeStr = "";
+            switch (mode) {
+                case GREE_MODE_AUTO:
+                    modeStr = MODE_AUTO;
+                    break;
+                case GREE_MODE_COOL:
+                    boolean powerSave = device.getIntStatusVal(GREE_PROP_PWR_SAVING) == 1;
+                    modeStr = !powerSave ? MODE_COOL : MODE_ECO;
+                    break;
+                case GREE_MODE_DRY:
+                    modeStr = MODE_DRY;
+                    break;
+                case GREE_MODE_FAN:
+                    modeStr = MODE_FAN;
+                    break;
+                case GREE_MODE_HEAT:
+                    modeStr = MODE_HEAT;
+                    break;
+                default:
+                    modeStr = String.valueOf(mode);
+
+            }
+            if (!modeStr.isEmpty()) {
+                logger.debug("{}: Updading mode channel with {}/{}", thingId, mode, modeStr);
+                return new StringType(modeStr);
+            }
+        }
+        return null;
+    }
+
+    private @Nullable State updateQuiet() throws GreeException {
+        if (device.hasStatusValChanged(GREE_PROP_QUIET)) {
+            switch (device.getIntStatusVal(GREE_PROP_QUIET)) {
+                case GREE_QUIET_OFF:
+                    return new StringType(QUIET_OFF);
+                case GREE_QUIET_AUTO:
+                    return new StringType(QUIET_AUTO);
+                case GREE_QUIET_QUIET:
+                    return new StringType(QUIET_QUIET);
+            }
+        }
+        return null;
+    }
+
+    private @Nullable State updateTemp() throws GreeException {
+        if (device.hasStatusValChanged(GREE_PROP_SETTEMP) || device.hasStatusValChanged(GREE_PROP_TEMPUNIT)) {
+            int unit = device.getIntStatusVal(GREE_PROP_TEMPUNIT);
+            return toQuantityType(device.getIntStatusVal(GREE_PROP_SETTEMP), DIGITS_TEMP,
+                    unit == TEMP_UNIT_CELSIUS ? SIUnits.CELSIUS : ImperialUnits.FAHRENHEIT);
+        }
+        return null;
+    }
+
+    private String logInfo(String msgKey, Object... arg) {
+        String message = messages.get(msgKey, arg);
+        logger.info("{}: {}", thingId, message);
+        return message;
+    }
+
+    public static QuantityType<?> toQuantityType(Number value, int digits, Unit<?> unit) {
+        BigDecimal bd = new BigDecimal(value.doubleValue());
+        return new QuantityType<>(bd.setScale(digits, BigDecimal.ROUND_HALF_EVEN), unit);
+    }
+
+    private void stopRefrestTask() {
+        forceRefresh = false;
+        if (refreshTask == null) {
+            return;
+        }
+        synchronized (refreshTask) {
+            ScheduledFuture<?> task = refreshTask;
+            if ((task != null) && !task.isCancelled()) {
```

Comment:
No harm in cancelling something that is already cancelled.
```suggestion
            if (task != null) {
```

Reply:
changed